torch/testing/_internal/common_utils.py

```diff
@@ -1049,6 +1049,40 @@ def setUp(self):
 
         set_rng_seed(SEED)
 
+    def genSparseCSRTensor(self, size, nnz, device, *, dtype, index_dtype):
+        sparse_dim = 2
+        assert all(size[d] > 0 for d in range(sparse_dim)) or nnz == 0, 'invalid arguments'
+
+        def random_sparse(m, n, nnz_per_row):
+            rows = torch.arange(m, dtype=index_dtype, device=device).repeat(nnz_per_row)
+            cols = torch.randint(0, n, size=[nnz_per_row * m], dtype=index_dtype, device=device)
+            return rows, cols
+
+        def coo_to_csr(indices, dim):
+            r = torch.zeros(dim + 1, dtype=index_dtype)
+            for i in range(nnz):
+                hp0 = indices[i]
+                hp1 = dim if i + 1 == nnz else indices[i + 1]
+                if hp0 != hp1:
+                    r[hp0 + 1:hp0 + 1 + hp1] = i + 1
+            return r
+
+        values = make_tensor([nnz], device=device, dtype=dtype, low=-1, high=1)
+        rows, cols = None, None
+        if nnz >= size[0]:
+            rows, cols = random_sparse(size[0], size[1], nnz // size[0])
+        else:
+            rows, cols = random_sparse(size[0], size[1], 1)
+            permutation = list(range(nnz))
+            random.shuffle(permutation)
+            rows = rows[permutation]
+            cols = cols[permutation]
+        crow_indices = coo_to_csr(rows, size[0])
+        col_indices = cols
+        return torch.sparse_csr_tensor(crow_indices,
+                                       col_indices,
+                                       values, size=size, dtype=dtype, device=device)
```

Comment:
```
[------------------------------------------------ sparse.mm -------------------------------------------------]
                          |  gen_sparse_csr  |  gen_sparse_csr_from_coo  |  gen_sparse_csr_from_gen_coo_sparse
1 threads: ---------------------------------------------------------------------------------------------------
      (32, 100) | 0.1     |       17.9       |            4083.4         |                 1092.2             
      (32, 100) | 0.25    |       21.6       |            9828.9         |                 2254.4             
      (512, 100) | 0.1    |       65.5       |           62794.7         |                16265.3             
      (512, 100) | 0.25   |      142.7       |          166035.6         |                37583.0             
      (1024, 100) | 0.1   |      121.4       |          140723.3         |                34030.6             
      (1024, 100) | 0.25  |      281.1       |          372286.9         |                75763.5             
      (2048, 100) | 0.1   |      219.9       |          257356.1         |                65103.0             
      (2048, 100) | 0.25  |      521.3       |          657451.4         |               147426.5      
```
I compared three methods using this script [https://gist.github.com/aocsa/9594f45b336065e4f887006110f0df2b](url)   `gen_sparse_csr  |  gen_sparse_csr_from_coo  |  gen_sparse_csr_from_gen_coo_sparse`, the winner is `gen_sparse_csr` which generate directly CSR components.
  
Note that `gen_sparse_csr_from_gen_coo_sparse` which uses `genSparseTensor` is less flexible because it can't change the `index_dtype`, which in the case of CSR could be int32 or int64 compared with COO which by default is int64.

Reply:
@aocsa thanks for these benchmarks.

Interestingly, generating CSR samples is more efficient than generating COO samples. For instance, defining
```
def gen_sparse_coo(size, nnz, dtype=torch.float, index_dtype=torch.int64, device='cpu'):
    return genSparseTensor(size, 2, nnz, False, device, dtype)[0]
```
and disabling the last two columns, we have the following benchmark results:
```
                          |  gen_sparse_csr  |  gen_sparse_coo
1 threads: ---------------------------------------------------
      (32, 100) | 0.1     |       20.6       |        75.0    
      (32, 100) | 0.25    |       26.6       |       113.6    
      (512, 100) | 0.1    |       80.0       |       492.0    
      (512, 100) | 0.25   |      173.1       |      1198.1    
      (1024, 100) | 0.1   |      142.6       |       965.8    
      (1024, 100) | 0.25  |      326.9       |      2357.4    
      (2048, 100) | 0.1   |      267.5       |      1868.4    
      (2048, 100) | 0.25  |      642.3       |      5802.0    
```
that is, sampling COO is about 4-9x slower than a sampling of CSR.

This may indicate that COO sampling could be made more efficient, or perhaps not. For instance, COO sampler uses `torch.rand->mul->to(long)` to compute `indices` while CSR uses just 1D `torch.randint`.